### ISSUE TITLE
web crash due to aborting during 'try' step followed by 'retry' step

### ISSUE THREAD

**Issue by hprotzek** [concourse/concourse#3989]
## Bug Report

We are running 3 web nodes on k8s (gke), after a rolling restart under high load, lots of running pipelines from 50 workers, the web nodes crashed multiple times with:

```
runtime error: invalid memory address or nil pointer dereference [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xaa585a]
at github.com/concourse/concourse/atc/exec.(*RetryStep).Succeeded (retry_step.go:47)
at github.com/concourse/concourse/atc/exec.OnSuccessStep.Succeeded (on_success.go:45)
at github.com/concourse/concourse/atc/exec.OnSuccessStep.Succeeded (on_success.go:45)
at github.com/concourse/concourse/atc/exec.OnFailureStep.Succeeded (on_failure.go:44)
at github.com/concourse/concourse/atc/engine.(*execBuild).Resume (engine.go:198)
```

## Steps to Reproduce

We couldn't reproduced it, but this was the second time we encountered this problem. 3 web nodes, 50 workers and lots of running pipelines.

## Version Info

* Concourse version: 5.2.0, 5.1.0
* Deployment type (BOSH/Docker/binary):  k8s
* Infrastructure/IaaS: gcp

**Comment by cirocosta** [comment]
Hi @hprotzek ,

Thanks for filling the issue!

What do you mean by `5.2.0,5.1.0`? Did you see the very same error happening on both versions?

@ddadlani given that the problem is at `exec`, is this something that the runtime team could look at?

Thanks!

**Comment by hprotzek** [comment]
@cirocosta yes we also saw it with version 5.1.0

**Comment by vito** [comment]
Might need this tracked down to a pipeline config that caused it in order to reproduce it. `Succeeded` should only be called after `Resume` was called so that panic is interesting. 🤔

**Comment by hprotzek** [comment]
We are using https://github.com/springernature/halfpipe to generate our pipelines. You can have a look at the examples: https://github.com/springernature/halfpipe/blob/master/e2e/artifacts/expected-pipeline.yml to see how most of our pipelines are structured.

**Comment by josecv** [comment]
We also encountered this in our concourse cluster, which is also deployed on kubernetes, and also has 3 web pods.

The sequence of events seems to have been as follows:

* Introduced a change to add `attempts: 5` to a number of resource `get`s and `put`s in a number of pipelines.
* After a little while, did a `kubectl edit deployment concourse-web` to change an environment variable, which rolling-restarted the web nodes.
* One of the web pods immediately went into CLBO (`CrashLoopBackOff`) with the same runtime error in the OP.
* After around 40 minutes, the other two web pods also started panicking and went into CLBO.
* We eventually killed all the workers (to forcefully abort any existing builds), killed all the web nodes, and walked back the `attempts: 5` change. Concourse has been stable since.

I'm also seeing `attempts: 2` in a number of resource `put`s in @hprotzek's pipeline, so I suspect that resource step retries are somehow related to this.

**Comment by misatos** [comment]
We've been able to repro it when we roll the web pods while many pipelines are in use
During downtime (e.g. at midnight) we were able to roll the pods without CLBO, but during the day, we've seen it CLBO thrice

During downtime (e.g. at midnight) we were able to roll the pods without CLBO
we did also observe a (related?) thing when we rolled the pods at midnight -- the pods were rolled, and were servicing web traffic, but pipelines weren't scheduling, and the web pods looked to be idling

i wonder if maybe, had lots of pipelines been live during that idle time the pods would've started CLBOing?

we never saw this idling when rolling concourse 4 web pods

**Comment by josecv** [comment]
I think I've found the underlying issue here. It doesn't actually seem connected to the web nodes restarting but rather to a given pipeline structure being aborted.

Here's a repro pipeline:

```
jobs:
  - name: rip-concourse
    plan:
      - try:
          task: try-task
          config:
            platform: linux
            image_resource:
              type: docker-image
              source:
                repository: alpine
                tag: 3.8
            run:
              path: /bin/sh
              args:
                - -c
                - echo try hitting that abort button && sleep 10
      - task: bye-concourse
        attempts: 1
        config:
          platform: linux
          image_resource:
            type: docker-image
            source:
              repository: alpine
              tag: 3.8
          run:
            path: /bin/sh
            args:
              - -c
              - echo i should be dead by now
```

Try aborting while `try-task` is running. It should crash the web pods.

I believe the root cause has to do with how `task_step.go` is structured. In its `Run` method, when the inner task returns an `abort` it is [propagated upward](https://github.com/concourse/concourse/blob/cb9e92d7eef49ed71baa2d180e7fb835f6ae37a3/atc/exec/try_step.go#L23-L26). `on_success` reads this error and [decides not to run the success hook](https://github.com/concourse/concourse/blob/cb9e92d7eef49ed71baa2d180e7fb835f6ae37a3/atc/exec/on_success.go#L30-L32). Which means that the `retrty_step`'s [`Run`](https://github.com/concourse/concourse/blob/5ffc88a8d5c7c82a225ab11997ddfc1fde230e18/atc/exec/retry_step.go#L22) method is never called. This means `LastAttempt` remains uninitialized.

Later on, when `engine.go` asks the `on_success` whether it was successful, `on_success` [delegates](https://github.com/concourse/concourse/blob/5ffc88a8d5c7c82a225ab11997ddfc1fde230e18/atc/exec/retry_step.go#L46-L48) to its inner step. In this case, the inner step is the try, which [always reports a success](https://github.com/concourse/concourse/blob/cb9e92d7eef49ed71baa2d180e7fb835f6ae37a3/atc/exec/try_step.go#L33). So `on_success` goes on to ask the hook whether it was successful, _even though the hook never ran at all!_ `retry_step` then goes on to [ask](https://github.com/concourse/concourse/blob/5ffc88a8d5c7c82a225ab11997ddfc1fde230e18/atc/exec/retry_step.go#L46-L48) its `LastAttempt` if it was successful aaand everything comes to a screeching halt, since `LastAttempt` is nil.

@vito what do you think is a good resolution here? It seems to me like the real problem is that `task_step` has a mismatch between what it considers a success and what it actually runs. It should probably keep some internal state to remember if it was aborted, and in the case that it was return `false` instead of `true`.

**Comment by cirocosta** [comment]
What a great reproducible, @josecv ! It's indeed a bomb aahah

Thank you so much for putting the work on exploring it and coming down to the core of it - thanks!

cc @vito and @pivotal-jwinters (the issue we talked about the other day)

**Comment by vito** [comment]
Yeah, great find! I'm actually pretty sure I've seen this panic before, but we never got to the bottom of it because we never saw the pipeline config that led to it. :+1: 

> @vito what do you think is a good resolution here? It seems to me like the real problem is that `try_step` has a mismatch between what it considers a success and what it actually runs. It should probably keep some internal state to remember if it was aborted, and that case return `false` instead of `true`. If that sounds like a good solution, I'd be happy to set up a PR to implement it.

This makes sense to me. Since `Try` already special-cases the 'abort' case by propagating its error, it makes sense to me that it would be aware of this in `Succeeded()`.

**Comment by vito** [comment]
I put this in our backlog so we'll get to it at some point, in case you don't have time to submit a PR. :)

**Comment by vito** [comment]
Also related: https://github.com/concourse/concourse/issues/3594

**Comment by josecv** [comment]
@vito I made the PR https://github.com/concourse/concourse/pull/4252 🙂

**Comment by vito** [comment]
merged #4252 - thanks!
